Pakfile Explorer fixes

tsa96 opened the pull request:
Fixes Panzer's issue with surf_arcade
Closes https://github.com/momentum-mod/lumper/issues/49
Closes #140 
Closes #133 
Closes #90

Panzerhandschuh commented on src/Lumper.UI/ViewModels/Shared/Pakfile/PakfileLumpViewModel.cs:
```diff
@@ -291,6 +291,22 @@ out string[]? opNoExtension
         }
+
+        // TexData - doesn't have viewmodels (yay!)
+        if (oldPath.EndsWith(".vmt", cmp) && opPrefix.Equals("materials", cmp))
```
Use `Path.GetExtension` instead.

tsa96 replied:
Coming back to this branch and refactoring a bunch of this method + moving it to model code. I don't think `Path.GetExtension` here is appropriate; this code does a lot of very Source-specific path refactoring, not using OS-dependent separators or the likes so `Path` methods aren't applicable in a lot of places. Given this I think it's much cleaner to only using general string methods, not a combination of string methods and Path methods.

GordiNoki commented on src/Lumper.Lib/BSP/Lumps/BspLumps/PakFileLump.cs:
```diff
@@ -299,17 +207,88 @@ private void UpdateZip()
+        string[] npSplit = newPath.Split(separator);
+
+        // VMTs can reference VTFs ignoring the root directory and without the extension
+        oldPath = string.Join(separator, opSplit[1..]);
```
This makes `oldPath` and `newPath` empty, making line 247 throw an error on string replace. If current path doesn't use specified separator, `opSplit` and `npSplit` would become one-element arrays, and both of them are sliced of here

tsa96 replied:
This exact code is very out-of-date now but I've added a check for paths with no separators in newer version of path refactoring.

Panzerhandschuh commented on src/Lumper.UI/ViewModels/Shared/Pakfile/PakfileLumpViewModel.cs:
```diff
@@ -210,86 +214,76 @@ PakfileEntryTextViewModel entry in Entries
+                // fucked.
+                if (directoryMatch is not null)
+                {
+                    op = string.Join("/", op.Split('/')[1..]);
```
It might be more efficient to use substrings:
`op = op.Substring(op.IndexOf("/") + 1);`

tsa96 replied:
Switched to using `Substring` where applicable in new version of pakfile refactoring

tsa96 commented:
Okay sorry, I merged a PR on top this one by mistake, and since realised the way I was doing pakfile refactoring was really misguided, doing some of it in the viewmodel code, some of it in the model. I've cleaned all that up significantly now, and fixed other related issues with pakfile refactoring

So this is essentially a new PR, Panzer's and Gordi's comments are still relevant to the new refactoring stuff though so using the existing PR.

Panzerhandschuh commented on src/Lumper.Lib/BSP/Lumps/BspLumps/PakFileLump.Refactoring.cs:
```diff
@@ -0,0 +1,334 @@
+    //   "$basetexture" "foo/bar"
+    // }
+    //
+    // Source will expects this file to be in materials/foo/bar.vtf.
+    // SourceRootDirectories lets our search omit the materials/ part,
+    // then IgnoreExtensions lets us omit the .vtf part.
+    //
+    // There's bound to be other file format combinations we could add here, I just
+    // havent't spent much time looking. When adding new entries be *very* careful -
+    // it make be that for some files, the cases where the extensions can be omitted
+    // depend on specific KV1 keys.
```
(nit) Typos in this comment

Panzerhandschuh commented on src/Lumper.Lib/BSP/Lumps/BspLumps/PakFileLump.Refactoring.cs:
```diff
@@ -0,0 +1,334 @@
+        return updated;
+    }
+
+    private bool UpdatePakfilePathReferences(
```
Does this handle updating sound paths in soundscape files?

Panzerhandschuh commented on src/Lumper.Lib/BSP/Lumps/BspLumps/PakFileLump.Refactoring.cs:
```diff
@@ -0,0 +1,334 @@
+        return count > 0;
+    }
+
+
```
(nit) Extra whitespace

tsa96 commented on src/Lumper.Lib/BSP/Lumps/BspLumps/PakFileLump.Refactoring.cs:
```diff
@@ -0,0 +1,334 @@
+            if (UpdateTexDataPathReferences(oldPath, newPath, opPrefix, opNoPrefix))
+                updatedTypes.Add(PathReferenceUpdateType.TexData);
+        }
+
```
Oh this was just me picking slightly the wrong lines when splitting up commits, not an issue in latest commit. Not gonna bother spend 5 mins tweaking the commits, not a big enough deal

expectchaos commented on src/Lumper.Lib/BSP/Lumps/BspLumps/PakFileLump.Refactoring.cs:
```diff
@@ -0,0 +1,357 @@
+            return false;
+
+        int match = pathList.FindIndex(name => name.StartsWith(oldPath, Comparison));
+        if (match != -1)
+            return false;
```
Needs to be if (match == -1)

expectchaos commented on src/Lumper.Lib/BSP/Lumps/BspLumps/PakFileLump.Refactoring.cs:
```diff
@@ -0,0 +1,357 @@
+                string cubemapName = match.Groups[1].Value;
+                entry.Rename(entry.Key.Replace(cubemapName, baseFilename));
+
+                entriesModified.Add((oldString, entry.Key));
```
Could use a Logger.Info for changed entries. Didn't see any logs when testing and wasn't sure if it was working.

expectchaos reviewed (commented):
Some broken or weird things I found while testing

tsa96 commented on src/Lumper.Lib/BSP/Lumps/BspLumps/PakFileLump.Refactoring.cs:
```diff
@@ -0,0 +1,357 @@
+        if (matched)
+        {
+            IsModified = true;
+            UpdateZip();
```
Ah yeah I forgot that UpdateZip was only needed in the uncompressed codepath, so fine to skip here.

tsa96 commented on src/Lumper.Lib/BSP/Lumps/BspLumps/PakFileLump.Refactoring.cs:
```diff
@@ -0,0 +1,357 @@
+    /// Renames the cubemap path as Source uses the filename when searching.
+    /// Returns a dictionary with the key as the old string and the value as the new string.
+    /// </summary>
+    public List<(string oldPath, string newPath)> RenameCubemapPaths(string newFileName)
```
Yeah agreed, new system moves everything

tsa96 commented:
Really appreciate the review @expectchaos! Since so many issues were coming up I've added support for unit testing and written ~70 tests for the pakfile refactoring stuff.

Usually I'd do that as a separate PR but wanted to be able to address all of Chaos' stuff. @Panzerhandschuh if you could review the latest stuff here that'd be great, includes the NUnit setup.
